scrnaseq.Rmd

```diff
@@ -1674,142 +1748,400 @@ if (degs_found) {
+
+```{r computing_deg_contrasts, warning=FALSE, results="hide"}
+# We first compute the DEGs for all requested contrasts
+
+# Prepare a list with contrasts (input can be R data.frame table or Excel file)
+degs_contrasts_list = DegsSetupContrastsList(sc, param$deg_contrasts, param$latent_vars)
+
+# Add the actual data to the list
+degs_contrasts_list = purrr::map(degs_contrasts_list, function(contrast){
+  # if there were already errors, just return
+  if (length(contrast[["error_messages"]])>0) return(c(contrast, list(object=NULL, cells_group1_idx_subset=as.integer(), cells_group2_idx_subset=as.integer())))
   
-      # Remember top term per geneset
-      for (j in seq(enriched[[i]])) {
-          enriched_top = rbind(enriched_top, 
-                               c(names(enriched)[i], 
-                               names(enriched[[i]])[j], 
-                               enriched[[i]][[j]][1, c("Term", "Overlap", "Adjusted.P.value", "Genes")]))
-      }
-    }
+  # cells indices
+  cells_group1_idx = contrast[["cells_group1_idx"]]
+  cells_group2_idx = contrast[["cells_group2_idx"]]
+
+  # object
+  if (contrast[["use_reduction"]]) {
+    # use dimensionality reduction
+    contrast[["object"]] = Seurat::Reductions(sc, slot=contrast[["assay"]])
+  } else {
+    # use assay
+    contrast[["object"]] = Seurat::GetAssay(sc[,unique(c(cells_group1_idx,cells_group2_idx))], assay=contrast[["assay"]])
+    
+    # This saves a lot of memory for parallelisation
+    if (contrast[["slot"]]!="scale.data") contrast[["object"]]@scale.data = new(Class = 'matrix')
   }
+  
+  # latent vars must be passed as data.frame
+  if (!is.null(contrast[["latent_vars"]]) && length(contrast[["latent_vars"]])>0) contrast[["latent_vars"]] = sc[[unique(c(cells_group1_idx,cells_group2_idx)), contrast[["latent_vars"]], drop=FALSE]]
+  
+  # now update cell indices so that they match to subset
+  contrast[["cells_group1_idx_subset"]] = match(colnames(sc)[cells_group1_idx], colnames(contrast[["object"]]))
+  contrast[["cells_group2_idx_subset"]] = match(colnames(sc)[cells_group2_idx], colnames(contrast[["object"]]))
+  
+  return(contrast)
+})
 
-  # Print table of top terms per gene set
-  knitr::kable(enriched_top, align="l", caption="Top enriched term per geneset") %>% 
-    kableExtra::kable_styling(bootstrap_options = c("striped", "hover")) %>% 
-    kableExtra::scroll_box(width="100%", height="700px")
-}
+# Compute the tests; TODO: this chunk may be done in parallel in future
+degs_contrasts_results = purrr::map(degs_contrasts_list, function(contrast) {
+  if (length(contrast$error_messages)==0) {
+    # No errors, do contrast
+    test_results = DegsTestCellSets(object=contrast[["object"]],
```

Comment:
Since DEG comparisons always take so long, the main idea here is to compute the DEG comparisons in parallel, i.e. in 8 threads in parallel. However, in worst case scenario depending on memory managment, R will then create a copy of `sc` for each thread which will make the memory requirement explode. 

Thus, we do the following:
- we keep only the assay that will be used for testing
- when the test is only done on a subset of cells, we keep only the corresponding subset of the assay
- if possible, we discard `scale.data` since while `counts` and `data` are sparse matrices and therefore do not take up much space, `scale.data` is actually a full matrix and contributes a lot to the size of the object